File: content/snippets/cpp/interview_lcci/17/22_word_transformer.cpp
#include <string>
#include <vector>
#include <unordered_set>
#include <queue>
using namespace std;

class Solution {
public:
    vector<string> findLadders(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> wordSet(wordList.begin(), wordList.end());
        if (!wordSet.count(endWord)) return {};

        queue<vector<string>> q;
        q.push({beginWord});
        unordered_set<string> visited = {beginWord};

        while (!q.empty()) {
            auto path = q.front(); q.pop();
            string cur = path.back();
            for (int i = 0; i < (int)cur.size(); ++i) {
                char orig = cur[i];
                for (char c = 'a'; c <= 'z'; ++c) {
                    if (c == orig) continue;
                    cur[i] = c;
                    auto newPath = path;
                    newPath.push_back(cur);
                    if (cur == endWord) return newPath;
                    if (wordSet.count(cur) && !visited.count(cur)) {
                        visited.insert(cur);
                        q.push(newPath);
                    }
                    cur[i] = orig;
                }
            }
        }
        return {};
    }
};

File: content/snippets/typescript/interview_lcci/17/22_word_transformer.ts
function findLadders(beginWord: string, endWord: string, wordList: string[]): string[] {
    const wordSet = new Set(wordList);
    if (!wordSet.has(endWord)) return [];

    const queue: string[][] = [[beginWord]];
    const visited = new Set([beginWord]);

    while (queue.length > 0) {
        const path = queue.shift()!;
        const cur = path[path.length - 1];
        const arr = cur.split('');
        for (let i = 0; i < arr.length; i++) {
            const orig = arr[i];
            for (let c = 0; c < 26; c++) {
                const ch = String.fromCharCode(97 + c);
                if (ch === orig) continue;
                arr[i] = ch;
                const nxt = arr.join('');
                if (nxt === endWord) return [...path, nxt];
                if (wordSet.has(nxt) && !visited.has(nxt)) {
                    visited.add(nxt);
                    queue.push([...path, nxt]);
                }
                arr[i] = orig;
            }
        }
    }
    return [];
}
